rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+*   **TF will become a suite of multiple packages, built by multiple authorities. What if the bugs get blamed on TF team**: With the modular model, we expect testing of 3rd party code to become easier. This can also be mitigated if the error messages are better, and if they can clearly point out which module the issue stems from. Finally, we can create an apple-swift like testing model, where we run a Jenkins setup that people can donate their machines to, and we can run continuous integration tests on their plugins.
+*   **Why not have APIs but still have a monolithic repository: **When everything is in a single repository, this enables developers to bypass the APIs, and depend on internals. Moreover, we cannot grant full control over different folders on our repository to our partners in a single repository. As long as they are in a single repository, they are still constrained by our build system and license. Finally, in a single repository we do not provide the option of closed source plugins for contributors.
+*   **Why not go with the OSS federation solutions?** OSS federation requires all dependencies to be in the federation before adding a repository. This is simply not possible for tensorflow, as eigen, llvm and many other dependencies will never be a part of the federation.
+*   **Documentation, how/where do we document everything?** With multiple repositories, structure of the documentation will need to be rethought, based on what is a part of "TensorFlow proper" and what is an optional feature.
+
+
+## Testing Plan
+
+We propose the following principles to be followed for testing in a modular world:
+
+* Each plugin tested separately.
+* Modules can plan their own integration tests.
+* Cut combinatorial explosion by divide and conquer.
+* Fuzzing at the core-module interface level if possible, in the case we need data marshalling between layers.
+* With this proposal, we aim to also simplify testing of tensorflow code. The biggest gain we expect in the modular world will be, we will be able to "divide and conquer" the TensorFlow support matrix.
+* Following describes an early modularized tensorflow package structure:
+
+
+![alt_text](20190305-modular-tensorflow/initial_tf_deps.png "Initial package structure")
+
+In the current setup, we need to test all of the above packages for different Python versions, operating systems, accelerators (CPU, GPU), compilers, and more variants combined. In the modularized world, each of these packages only need to be unit tested for the following:
+
+
+*   tensorflow-base: Operating systems, compiler versions and python versions only with CPU 
+*   tf-gpu: With GPU only, for different operating systems.
+*   tf-estimator: Only for different python versions
```

Comment:
Who will own publication of Tensorflow + MKL binaries?
- anaconda packages
- pypi packages
- container images

Reply:
Just describing from the current state (as official builds do not have MKL enabled yet), TF team releases main TF package, and Intel releases tf-mkl plugin. Again,  this is reflecting only the current state, and not taking into account any potential changes to mkl support in TF. We can evaluate those changes orthogonally. To clarify the 3 points you raised in the above described scenario:

- Anaconda packages: TF team does not own any anaconda packages. We are separately evaluating anaconda, so it is an orthogonal issue.
- pypi packages: `tensorflow` pypi package is owned by TF team, and `tensorflow-mkl` is a plugin owned by Intel.
- container images: same as pypi packages, but Intel has the option to extend the TF container images just with `pip install tensorflow-mkl` command to build the MKL images on top of TF images.

Reply:
CC @agramesh1